src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -712,8 +715,70 @@ static int get_pool_offset(sd_dhcp_server *server, be32_t requested_ip) {
         return be32toh(requested_ip & ~server->netmask) - server->pool_offset;
 }
 
-static int dhcp_server_relay_message(sd_dhcp_server *server, DHCPMessage *message, size_t opt_length) {
+static size_t relay_option_length(sd_dhcp_server *server) {
+        size_t length = 0;
+
+        assert(server);
+
+        if (server->agent_circuit_id)
+                length += strlen(server->agent_circuit_id) + 2;
+        if (server->agent_remote_id)
+                length += strlen(server->agent_remote_id) + 2;
+
+        return length;
+}
+
+static uint8_t *append_suboption(uint8_t *dst, size_t dst_len, uint8_t subcode, const char *value, size_t length) {
+        assert(dst);
+        assert(value);
+
+        if (dst_len < length + 2)
+                return NULL;
+        dst[0] = subcode;
+        dst[1] = length;
+        memcpy(dst + 2, value, length);
+        return dst + 2 + length;
+}
+
+static int append_agent_information_option(sd_dhcp_server *server, DHCPMessage *message, size_t msg_len, size_t buf_len) {
+        assert(server);
+        assert(message);
+
+        int r;
+        uint8_t inf[255];
+        const uint8_t *inf_end = inf + sizeof(inf);
+        uint8_t *dst = inf;
+        uint8_t *end_option_ptr = memrchr(message->options, SD_DHCP_OPTION_END, msg_len - sizeof(DHCPMessage));
+        if (!end_option_ptr)
+                return -EINVAL;
+        size_t options_length = end_option_ptr - message->options;
```

Comment:
Please declare variables at first, then assert.

And, please do not double copy the strings. How about the following:
```
uint8_t *end_option_ptr;
size_t offset, l;
int r;

assert(server);
assert(message);

l = relay_option_length(server);
if (l == 0)
        return 0;
if (l > 255)
        return log_dhcp_server_errno(server, SYNTHETIC_ERRNO(ENOBUFS), "relay agent circuit and/or remote id is too long: %m");

end_option_ptr = memrchr(...);
if (!end_option_ptr)
        return -EBADMSG;

offset = end_option_ptr - message->options + 2;

if (server->agent_circuit_id) {
        r = option_append(message->options, buf_len - sizeof(DHCPMessage), &offset, SD_DHCP_RELAY_AGENT_CIRCUIT_ID, strlen(server->agent_circuit_id), server->agent_circuit_id);
        if (r < 0)
                return r;
}
if (server->agent_remote_id) {
        r = option_append(message->options, buf_len - sizeof(DHCPMessage), &offset, SD_DHCP_RELAY_AGENT_REMOTE_ID, strlen(server->agent_remote_id), server->agent_remote_id);
        if (r < 0)
                return r;
}

end_option_ptr[0] = SD_DHCP_OPTION_RELAY_AGENT_INFORMATION;
end_option_ptr[1] = l;

r = option_append(message->options, buf_len - sizeof(DHCPMessage), &offset, SD_DHCP_OPTION_END, 0, NULL);
if (r < 0)
        return r;
```

Reply:
addressed in option_append call fix